### PR TITLE
`use_template()` no longer errors when a user chooses not to overwrite an existing file and instead produces a message

### LINKED ISSUE #348: Allow user to decline to overwrite existing file, from use_template()
Currently `use_template` throws an error when one chooses not to overwrite an existing file:
 
```
usethis::use_template(template = "CONTRIBUTING.md")
Overwrite pre-existing file 'CONTRIBUTING.md'?
1: Absolutely
2: No
3: Not now
 
Selection: 3
Error: 'CONTRIBUTING.md' already exists.
```
Is there any appetite for a PR that would add a `allow_existing` argument (or whatever the name) to `use_template`? So that is, if a file already exists and the `allow_existing` argument is set to `TRUE`, and a user selects to not overwrite (at the prompt), `use_template` simply outputs a 'CONTRIBUTING.md' already exists' warning rather than an error?
 
The context is that we have a series of templates (CoC, README, etc) where a user, when setting up a project or package, may accept the template for some and keep existing files for others. Since `use_template` via `write_over` throws an error, this [type of process where sequential asking which files need to be overwritten](https://github.com/bcgov/bcgovr/issues/37
)  is currently not possible.

### PR DESCRIPTION
**Pull request by boshek**
Closes #348 Allow user to decline to overwrite existing file, from use_template()

- Modified `write_over` so that it produces a message rather than an error when `can_overwrite()` is `FALSE`
- Revised relevant tests to test for a message rather than an error
- Added a single entry in the `NEWS` file outlining this small change (possibly not big enough to merit a NEWS item?)

### CODE REVIEW

**Comment by jennybc** [comment]
It looks like I'll be doing a small release of usethis soon (e.g. within the month), prompted by the backports update, so I will sort out these things once I dig in.

File: R/helpers.R [hunk 1]
```diff
@@ -45,7 +45,7 @@ use_template <- function(template,
     use_build_ignore(save_as)
   }
 
-  if (open) {
+  if (open & new) {
```
Comment (jennybc) [inline R/helpers.R]: This should be `&&`.

**Comment by jennybc** [comment]
Upon further thought, the failure on Travis re: `use_readme_rmd()` is real and needs to be dealt with.

Previously, we would have stopped inside `use_template()` if we can't overwrite an existing `README.Rmd`. But now we just message and go on to execute the next block, which calls `uses_git()` to decide whether to add a hook.

In general, you need to do a survey of all existing uses of `use_template()` and see if subsequent code assumes that a new template file was created. How to deal with this? It depends on how many functions are affected. Can you characterize the scope of this problem?

**Comment by boshek** [comment]
> Next time you probably be better off to make the PR from a non-master branch in your fork

Thanks for the tip.

Happy to characterize the scope of this. Hope to have some time near the end of this week to spend on this. Will that timing work?

**Comment by jennybc** [comment]
> Happy to characterize the scope of this. Hope to have some time near the end of this week to spend on this. Will that timing work?

No worries, I might get to it sooner or not. But we have to sort it out before we can merge this. As you might see, I now think it's connected to another issue #178.

**Comment by ateucher** [comment]
I hope you don't mind me piping in here... With this PR, `use_template()` (via `write_over()`) returns `FALSE` with a message now instead of failing if `can_overwrite` is `FALSE` … but that behaviour existed already [when the old file and the new file were identical](https://github.com/boshek/usethis/blob/a458e9318b7cb89dbf6b037ebf1e5801fed227df/R/write.R#L31-L32). So is this just exposing something that was previously untested?

So the question is (specifically for `use_readme_rmd()`) would you want the hook to be added if the `README.Rmd` already existed and wasn't overwritten? The existing behaviour would add the hook _if the file existed already but was identical to the template_. Is the new case really any different?

**Comment by ateucher** [comment]
This is some heavy yak-shaving, and possibly not very helpful, but here are the functions that call `use_template()` which have meaningful code that comes _after _the call to `use_template()`.

Most of them are just `todo()`/`code_block()` messages, except for:

- `use_travis()` and `use_appveyor()` call their respective `*_activate()` and `use_*_badge() functions.
- `use_readme_rmd()` we know about
- `use_rstudio()` calls `use_gitignore()`, and `use_build_ignore()` if it's a package.
- `use_tidy_ci()` and `use_rmarkdown_template()` call `use_template()` multiple times.

``` r
library(tidyverse)
library(igraph, quietly = TRUE, warn.conflicts = FALSE)
library(usethis)
# devtools::install_github("ropenscilabs/revtools")

find_code_after_use_template <- function(x) {
  x <- getFromNamespace(x, "usethis")
  fn_body <- body(x)
# find where use_template is called
  ut_block_no <- grep("use_template", fn_body)
  if (length(ut_block_no) > 1) {
    return("More than one call to use_template. Check this one out more carefully")
  }
# if call to use_template is the last line, or the line after the call is a return-type statement
  if (ut_block_no == length(fn_body) || grepl("^return|^invisible\\(", fn_body[ut_block_no + 1])) {
    return(character(0))
  }
# return the rest of the function body
  as.character(fn_body[seq(ut_block_no + 1, length(fn_body))])
}

# Get character vector of exported function names that call use_template
funs_that_call_use_template <- revtools:::create_package_igraph("~/dev/packages/usethis/") %>% 
  as_long_data_frame() %>% 
  filter(to_name == "use_template") %>% 
  pull(from_name)

map(funs_that_call_use_template, find_code_after_use_template) %>% 
  set_names(funs_that_call_use_template) %>% 
  Filter(function(x) length(x), .)
#> $use_travis
#> [1] "travis_activate(browse)" "use_travis_badge()"     
#> [3] "invisible(TRUE)"        
#> 
#> $use_appveyor
#> [1] "appveyor_activate(browse)" "use_appveyor_badge()"     
#> [3] "invisible(TRUE)"          
#> 
#> $use_code_of_conduct
#> [1] "todo(\"Don't forget to describe the code of conduct in your README.md:\")"                                                                                                                 
#> [2] "code_block(\"Please note that this project is released with a [Contributor Code of Conduct](CODE_OF_CONDUCT.md).\", \"By participating in this project you agree to abide by its terms.\")"
#> 
#> $use_pipe
#> [1] "todo(\"Run \", code(\"document()\"))"
#> 
#> $use_readme_rmd
#> [1] "if (uses_git()) {\n    use_git_hook(\"pre-commit\", render_template(\"readme-rmd-pre-commit.sh\"))\n}"
#> [2] "invisible(TRUE)"                                                                                      
#> 
#> $use_revdep
#> [1] "todo(\"Run checks with \", code(\"revdepcheck::revdep_check(num_workers = 4)\"))"
#> 
#> $use_rmarkdown_template
#> [1] "More than one call to use_template. Check this one out more carefully"
#> 
#> $use_rstudio
#> [1] "use_git_ignore(\".Rproj.user\")"                                                                     
#> [2] "if (is_package()) {\n    use_build_ignore(c(paste0(project_name(), \".Rproj\"), \".Rproj.user\"))\n}"
#> [3] "invisible(TRUE)"                                                                                     
#> 
#> $use_tidy_ci
#> [1] "More than one call to use_template. Check this one out more carefully"
#> 
#> $use_tidy_eval
#> [1] "todo(\"Run document()\")"
#> 
#> $use_tidy_coc
#> [1] "todo(\"Don't forget to describe the code of conduct in your README.md:\")"                                                                                                                         
#> [2] "code_block(\"Please note that this project is released with a [Contributor Code of Conduct](.github/CODE_OF_CONDUCT.md).\", \"By participating in this project you agree to abide by its terms.\")"
```

Created on 2018-05-30 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by boshek** [comment]
Using @ateucher's handy scoping:

### use_travis and use_appveyor
A user with an existing *.yml file here can safely choose not to overwrite because neither the active and badge functions depend on use_template failing when an overwrite does not occur

### use_readme_rmd
I think the question here identified by @ateucher is:
> would you want the hook to be added if the README.Rmd already existed and wasn't overwritten
I think you would right? 

I have skipped the readme_rmd test if R is less than 3.1 because it was trying to use `git2r::::discover_repository()` in the test now which fails on travis. Hopefully this fixes the build fail on 3.1.

### use_rstudio
The main impact I see here is that calling `create_project()` or `create_package()` from a directory that has an existing `.Rproj` file will now ask you if you want to overwrite it. 

This seems a little redundant so one could only call `use_template()` if an .Rproj is absent. Would this be desirable?

### use_rmarkdown_template & use_tidy_ci. 
This functions without any adverse downstream effects and actually provides greater control if one wants to replace the .yaml but not the rmarkdown template (or vice versa).

**Comment by boshek** [comment]
@jennybc Curious is it generally considered my responsibility to fix the merge conflicts that have arisen since the PR was originally submitted or is that usually left to the maintainer (you in this case)?

**Comment by jennybc** [comment]
You can leave it with me @boshek.

**Comment by jennybc** [comment]
@boshek I'm working on this now and will try to finish things up in your PR. This means I'm going to be pushing to the branch you made it from (and have already done so). In future, you probably want to work in a non-master branch for PRs, but we'll just carry on with this.

Are OK to just let me take over master branch of your fork for a little while?

**Comment by boshek** [comment]
@jennybc Fine by me. 

- I could, alternatively, close everything down and re-submit a clean PR if that were the path of least resistance at this point. 
- Happy to further contribute as well if any tasks can be farmed out

**Comment by jennybc** [comment]
@boshek I *think* this is done. I'm just starting a vacation would like to wrap up this release before I do a complete mental check out.

Will you sanity check this again, both by taking a look at the code I pushed and by confirming it still behaves the way you want? I'll be working through my other pre-release checklist items in parallel.

Thanks!

**Comment by jennybc** [comment]
@ateucher Review by you, instead or in addition, is also welcome!

**Comment by jennybc** [comment]
Thanks for taking a look!

> Note: It isn't totally clear to me how this, this and this commit aren't in the boshek-master branch but are still part of the PR here. I assume this reflect my lack of understanding but am flagging this just in case.

That's a branch in this repo that I created by mistake, in some sense. I will delete it. The branch that matters for the PR is the `master` branch in your fork, which does have these commits.

**Pull request merged**